Add support for CiliumEnvoyConfig CRD

michi-covalent opened the pull request:
```release-note
Add support for CiliumEnvoyConfig CRD.
```

jrajahalme committed:
k8s: Add CiliumEnvoyConfig CRD

Add a new CiliumEnvoyConfig to cilium.io API v2alpha1

Supported Envoy resource types are limited to Listener, Route, Cluster,
Endpoint, and Secret resources.

Support for Secret Discovery Service is not yet complete:
- Cilium xds cluster details need to be given on each instance,
  omitting SDS config source is not yet supported.
- Secrets are not fetched from k8s yet.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Add support for L7 load balancers

Add new SVC_FLAG_L7LOADBALANCER to designate services that are not
loadbalanced in BPF, but are forwarded to a local TPROXY loadbalancer
instead.

The set of proxy ports registered to datapath was limited to the
preconfigured set of HTTP, DNS, and proxylib ports. Allow adding (CRD)
proxy ports to this set.

Redirect L7 LB nodeport service connections received by the host to the
Envoy listener.

Redirect to nodeport L7 LB from bpf_sock for hostns sources. This is
needed due to this case not being handled by eBPF datapath on TC
level. Note that this requires a recent kernel to work (Linux
5.12+). With older kernels nodeport services only work between nodes,
not from a node host namespace to the node itself. This is fine for
Ingress functionality.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
service: Push Envoy Endpoints for L7 LB k8s services

Upsert Envoy endpoint resources for services that have been selected
for L7 LB in CiliumEnvoyConfig CRDs.

Rearrange daemon init order so that we can pass 'd.l7Proxy' to
NewService() as a new parameter. This way we can remove the Envoy stubs
from the daemon.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Add xDS resource validation

Make sure Envoy xDS resources have names and validate xDS resources
before adding them to xDS cache. This validation can not be done at
the k8s CRD apply time, so k8s CRD can contain invalid Envoy
configurations.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Allocate CEC listener ports if unspecified

Add support for dynamically created (CEC) proxy ports in pkg/proxy.

Allocate proxy ports for CEC listeners without Listener.Address
specification and install datapath TPROXY rules if
Listener.Address is left empty.

Allocated proxy ports are sticky in order to return the same port
when the same listener (by name) is reconfigured or configured again
after deletion.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
service: Register k8s service redirection to Envoy listener

Register L7 LB service redirections in the service package. Use
'L7LBInfo' to keep track of the CiliumEnvoyConfig that requested a
service to be registered for L7 load balancing. Only one
CiliumEnvoyConfig may request redirection of a given service at a time,
as otherwise the choice between Envoy listeners would be ambiguous.

Service backends are synced to Envoy, so that the ingress service can
load balance to them. Datapath L7LB redirection is only requested for
'services' in a CiliumEnvoyResources. Backends are synced to Envoy
also for 'backendServices'. This allows the regular eBPF LB to be used
for pod-to-pod traffic for 'backendServices' while L7 LB is used for
ingress to those same services.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Add support for re-entering LXC egress path after L7 LB

Allow host to tail call to lxc egress for policy enforcement after L7 LB.

This requires L7 LB proxy to set the source Endpoint ID in the socket
mark. This mark includes a new magic value that the datapath
recognizes to subject the traffic to source LXC (policy enforcement)
processing again. Note that this does not loop endlessly due to the L7
LB switching from service frontend IP to backend IP.

iptables rules are amended to avoid overwriting the new "from L7 LB"
mark with "from host" mark. L7 load balancer upstream connections may
go to the egress proxy listener of the source pod. Pod connections to
proxy listeners are already not tracked to avoid conflicts due to
reuse of the same 5-tuple on both sides of the (egress) proxy. Add
NOTRACK rules for L7 LB connections that may go to the egress proxy
for the same reason.

A new flag 'from_l7lb' is added to conntrack to enable redirecting
reply packets back to L7 LB proxy.

Add a new map for bpf_host to tail call into for LXC policy
enforcement after L7 LB.

Some kernels reached their eBPF complexity limit with the new code
compiled in. Address this by making all L7 LB related code in
bpf_lxc.c confitional on the new ENABLE_L7_LB. This allows the new
feature to be used on kernels that allow a but more complex eBPF
programs.

Checks for the need to tail call back to the source pod also in to
to-container and to-netdev programs, for traffic routed to a local or
remote pod, respectively.

Do not reuse external original source address with L7 LB. This fixes a
problem with Ingress that caused return traffic from remote-node backends
to be routed directly to the external source instead of coming back to
the Envoy listener.

Note that this causes external traffic redirected via Ingress to look
like traffic from local node to local backends, and traffic from
remote-node to remote backends. This undoubtedly complicates ingress
policy definition for the backends.

handle_policy_egress() is tail called for traffic coming from the host
proxy, so CB_IFINDEX should be zero. Assume this is the case.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Do not wait for RDS, CDS, EDS, or SDS

Do not wait for routes (RDS), clusters (CDS), cluster load assignments
(EDS), or secrets (SDS) as they will only be acked if referenced from
a listener that may not exist yet or may have been deleted already.

Add Cilium agent option --envoy-config-timeout to limit the maximum
time the agent waits for Envoy (Listener) configuration to be acked,
defaulting to 2 minutes. This unblocks k8s watchers if the agent ever
waits for Envoy for resources that are not being acknoowledged.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Inject Cilium policy enforcement filters to L7 LB config

Inject Cilium policy enforcement filters for CiliumEnvoyConfigs that do
not explicitly specify the listener port number. This way custom
listeners that would not work with policy enforcement can still be added,
while in the typical case Cilium network policy is still enforced.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras commented on bpf/bpf_sock.c:
```diff
@@ -982,11 +1013,20 @@ static __always_inline int __sock6_xlate_fwd(struct bpf_sock_addr *ctx,
+	/* See __sock4_xlate_fwd for commentary. */
+	if (lb6_svc_is_l7loadbalancer(svc)) {
+		if (is_defined(BPF_HAVE_NETNS_COOKIE) && in_hostns) {
+			l7backend.address = { .addr[15] = 1, };
```
Trying to set enable-envoy-config as true and check for IPv6 environment (e.g. kind conformance ipv6 test), but got the below error.

```
unknown-warning-option -Wno-gnu-variable-sized-type-not-at-end -Wdeclaration-after-statement -Wimplicit-int-conversion -Wenum-conversion -I. -I/run/cilium/state/globals -I/var/lib/cilium/bpf -I/var/lib/cilium/bpf/include -D__NR_CPUS__=32 -DENABLE_ARP_RESPONDER=1 -DCALLS_MAP=cilium_calls_lb -c /var/lib/cilium/bpf/bpf_sock.c -o -" subsys=datapath-loader
level=warning msg="+ llc -march=bpf -mcpu=v3 -mattr=dwarfris -filetype=obj -o bpf_sock.o" subsys=datapath-loader
level=warning msg="/var/lib/cilium/bpf/bpf_sock.c:1020:24: error: expected expression" subsys=datapath-loader
level=warning msg="                        l7backend.address = { .addr[15] = 1, };" subsys=datapath-loader
level=warning msg="                                            ^" subsys=datapath-loader
level=warning msg="1 error generated." subsys=datapath-loader
level=fatal msg="Error while creating daemon" error="error while initializing daemon: exit status 1" subsys=daemon

```

jrajahalme replied:
@sayboras Thanks for the report, I pushed a new commit that fixes this.

michi-covalent commented on pkg/envoy/ciliumenvoyconfig_test.go:
```diff
@@ -0,0 +1,145 @@
+	"sigs.k8s.io/yaml"
+
+	envoy_config_http "github.com/cilium/proxy/go/envoy/extensions/filters/network/http_connection_manager/v3"
+
```
minor nit: remove extra blank lines

jrajahalme replied:
Done.

michi-covalent commented on pkg/envoy/ciliumenvoyconfig.go:
```diff
@@ -0,0 +1,188 @@
+		log.Debug("UpdateEnvoyResources: Waiting for proxy deletes to complete...")
+		err := wg.Wait()
+		if err != nil {
+			log.Debug("UpdateEnvoyResources: delete failed: ", err)
```
should we return `err` if delete fails, or is this error safe to ignore?

jrajahalme replied:
Envoy likely fails to configure the listener with a new port number if the delete of the old listener here fails. This will then result in an error return from the later `wg.Wait()` in the same function. It could also be that the listener delete failure will not cause the new listener to fail. Either way it will be informative to see if any `UpdateEnvoyResources: delete failed:` lines appear in the logs.

Normally we let Cilium pick up the proxy ports, in which case the port will not change between reconfigurations, and we do not need to wait for the deletes to complete (`waitForDelete` remains `false`).

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1492,9 +1625,9 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
+#if defined(ENABLE_PER_PACKET_LB) && !defined(DISABLE_LOOPBACK_LB)
 skip_policy_enforcement:
-#endif /* !ENABLE_HOST_SERVICES_FULL && !DISABLE_LOOPBACK_LB */
+#endif /* ENABLE_PER_PACKET_LB && !DISABLE_LOOPBACK_LB */
```
Why is this change needed? It's subtly tucked into the commit about allowing re-injection of L7LB packets back into the egress path. Rather than extending the checks to include something about `L7LB`, it seems to be changing between `ENABLE_HOST_SERVICE_FULL` and `ENABLE_PER_PACKET_LB`, which seems like it could end up somehow impacting the datapath when L7LB is administratively disabled.

jrajahalme replied:
I separated this change to a commit of it's own:
```
    bpf: Fix conditional compilation

    Use the same condition in defining the label 'skip_policy_enforcement' as
    for jumping to it. Otherwise we could jump to a nonexisting label given a
    suitable datapath configuration. This does not seem possible now, as
    ENABLE_HOST_SERVICES_FULL should not be defined if
    ENABLE_SOCKET_LB_HOST_ONLY is defined. However, as the conditions for
    defining ENABLE_PER_PACKET_LB evolve, this would become a compilation
    bug.

Fixes: #17154
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
---
 bpf/bpf_lxc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/bpf/bpf_lxc.c b/bpf/bpf_lxc.c
index 7c52f9ec40..a9448b54ed 100644
--- a/bpf/bpf_lxc.c
+++ b/bpf/bpf_lxc.c
@@ -1517,9 +1517,9 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 					   verdict, policy_match_type, audited);
 	}
 
-#if !defined(ENABLE_HOST_SERVICES_FULL) && !defined(DISABLE_LOOPBACK_LB)
+#if defined(ENABLE_PER_PACKET_LB) && !defined(DISABLE_LOOPBACK_LB)
 skip_policy_enforcement:
-#endif /* !ENABLE_HOST_SERVICES_FULL && !DISABLE_LOOPBACK_LB */
+#endif /* ENABLE_PER_PACKET_LB && !DISABLE_LOOPBACK_LB */
 
 #ifdef ENABLE_NODEPORT
 	if (ret == CT_NEW || ret == CT_REOPENED) {
```

jrajahalme replied:
Separated these generic changes to a separate PR: #19104

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1564,7 +1697,7 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 	if (ifindex)
 		return redirect_ep(ctx, ifindex, from_host);
-#endif /* ENABLE_ROUTING && TUNNEL_MODE && !ENABLE_NODEPORT */
+#endif /* !ENABLE_ROUTING && TUNNEL_MODE && !ENABLE_NODEPORT */
```
nit: Would be nice to include tidyups like this separately so they aren't caught up in this specific patch's feature logic.

jrajahalme replied:
Thanks! I separated this to a separate commit.

jrajahalme replied:
Unfortunately chopping these off to an earlier commit means that I'll have to force push this PR, which will mess with some of these comments :-(

michi-covalent commented on pkg/k8s/apis/cilium.io/v2alpha1/cec_types.go:
```diff
@@ -0,0 +1,162 @@
+
+// +kubebuilder:pruning:PreserveUnknownFields
+type XDSResource struct {
+	*anypb.Any `json:"-"`
```
nit: could you add a comment here which xds resource can be specified here? i'm guessing it can be any resource in https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol?

jrajahalme replied:
Added this comment:
```
// XDSResource can carry the following Envoy resource types:
// - type.googleapis.com/envoy.config.listener.v3.Listener
// - type.googleapis.com/envoy.config.route.v3.RouteConfiguration
// - type.googleapis.com/envoy.config.cluster.v3.Cluster
// - type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment
// - type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret
//
// +kubebuilder:pruning:PreserveUnknownFields
type XDSResource struct {
	*anypb.Any `json:"-"`
}
```

jrajahalme replied:
I had to move this comment into `CiliumEnvoyConfigSpec.Resources` so that this info would appear in the CRD schema description of that field.

michi-covalent reviewed (commented):
:heavy_check_mark: reviewed https://github.com/cilium/cilium/pull/18894/commits/be59c6141c820d691ea012e31d367978123b5bfa looks good to me, just a couple of minor nits / questions.

joestringer commented on bpf/lib/common.h:
```diff
@@ -599,6 +603,12 @@ enum {
 #define	CB_CUSTOM_CALLS		CB_CT_STATE	/* Alias, non-overlapping */
 };
 
+/* Magic values for CB_FROM_HOST.
+ * CB_FROM_HOST overlaps with CB_NAT46_STATE, so this value must be distinct
+ * from any in enum NAT46 below!
+ */
+#define FROM_HOST_L7_LB 0xFACADE42
```
:thinking: all the other `CB_FROM_HOST` cases just set 1, which overlaps with the enum below. Why is this different? This commit ensures that the CB_FROM_HOST is populated with this one value then passes the packet to a hook that only interprets this particular CB field as the `CB_FROM_HOST`. I don't think there's anything unique that the L7LB code wants to do semantically with this flag, it just needs to pass down the fact that "this packet is coming from the host, therefore when deciding how to route the packet next, we need to decide whether we can redirect it or just let it through so that the stack routes it".

That said, I don't necessarily think that this is a reason to avoid this approach - maybe we should go even further, by putting all of these values clearly in the same enum so that we avoid ever making them overlap? But this is probably more of a longer term maintainability / readability question rather than needing to look at this closer in context of this PR.

joestringer replied:
Ah, in my initial scan I saw the cases where the semantics of this were the same as `CB_FROM_HOST`, but now that I see it getting passed into `ct_create[46]()`, I suspect it also has CT state storage semantics which are different from just identifying the traffic as "from the host" in order to route it correctly.

jrajahalme replied:
Yes, FROM_HOST_L7_LB value in CB_FROM_HOST is used to pass the fact that the traffic originated from L7 LB through the tail call back to the source lxc.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1742,6 +1874,61 @@ int handle_policy(struct __ctx_buff *ctx)
+	}
+
+	ifindex = ctx_load_meta(ctx, CB_IFINDEX);
+	ctx_store_meta(ctx, CB_IFINDEX, 0);
```
If this is coming from the host, then I don't think there's any ifindex. CB should probably already be clear. Can probably skip these two lines and hardcode the ifindex parameter below to `0`?

(I could also see an argument for figuring out the actual ifindex of the device associated with the specific LXC for this tail call, but at a glance this doesn't seem to be directly available).

jrajahalme replied:
I was not aware of this nuance so played it safe. Will try without. (added a separate commit we can squash in before merging)

jrajahalme replied:
Now removed, still works in my local testing.

joestringer commented on pkg/datapath/linux/linux_defaults/mark.go:
```diff
@@ -46,6 +50,18 @@ const (
+	// proxy which is capturing traffic before it is subject to egress
+	// policy enforcement that must be done after the proxy. The identity
+	// stored in the mark is source Endpoint ID.
+	MagicMarkEgressEPID int = 0x0900
```
These appear to be unused.

joestringer replied:
^^ Unless I'm missing something obvious on the iptables configuration side, IMHO ideally we'd name this variable something like `MarkMagicProxyEgressEPID` to keep it consistent with the corresponding `MARK_MAGIC_PROXY_EGRESS_EPID`. And of course consistent value like it has, and use the same value in the iptables rules so that it's clear how these disparate bits of Linux configuration relate to each other.

jrajahalme replied:
Changed the name as suggested. Also removed definition of `MagicMarkIngressEPID` as that pattern is actually not used (L7 LB only makes sense for egress).

Also added a comment that the definition of `MagicMarkProxyEgressEPID` is included here for documentation purposes and that this definition must match `MARK_MAGIC_PROXY_EGRESS_EPID` in datapath and the code in [cilium/proxy/cilium/bpf_metadata.cc](github.com/cilium/proxy/cilium/bpf_metadata.cc).

joestringer commented on pkg/datapath/linux/linux_defaults/mark.go:
```diff
@@ -37,6 +37,10 @@ const (
+	// MagicMarkIsProxyEPID can be used in conjunction with MagicMarkProxyMask
+	// to determine whether the mark is indicating that traffic is sourced
+	// from a proxy prior to endpoint policy enforcement.
+	MagicMarkIsProxyEPID int = 0x0800
```
Why is this set to `0x0800` while the `MARK_MAGIC_PROXY_EGRESS_EPID` is set to `0x0900`?

joestringer replied:
Related, the choice of `0x0900` on the datapath means that binary bits `1001` are set in that nibble, and the `MagicMarkProxyMask` already sets `0x0E00`, meaning that it also covers binary bits `1110`, so now the actual proxy-relevant bits in the mask are `0x0F00`. Maybe we're trying to avoid that lowest bit for some reason, eg because it's has some specific other meaning?

jrajahalme replied:
The lowest bit is already used with 0x0A00 and 0x0B00 to tell the difference between ingress/egress, where the lowest bit is 0 or 1, respectively.

I carried over the same semantic to "proxy EPID" mark, where ingress uses 0x0800 and egress uses 0x0900. For testing if the mark includes the endpoint ID, masking with 0x0800 is needed, so that the difference between ingress/egress is abstracted away.

Now it turns out that at least for now, we only need the "mark includes EPID" for egress. Use of these patterns is already built in to the Envoy image where the mark is set, but we could re-spin that and make both `MagicMarkProxyEgressEPID` and `MARK_MAGIC_PROXY_EGRESS_EPID` be 0x0800 instead. But then the semantic of the lowest bit in this nibble would be reversed from that of MagicMarkIsProxy...

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1742,6 +1874,61 @@ int handle_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+/* Handle policy decisions as the packet makes its way from the
+ * endpoint.  Previously, the packet has come from the same endpoint,
+ * but was redirected to a L7 LB.
+ *
+ * This program will be tail called from bpf_host for packets sent by
+ * a L7 LB.
+ */
```
How do we ensure that the L7LB'd packet arrives to the `cilium_host`/`cilium_net` veth pair? Do we statically configure routes for the service VIP range?

joestringer replied:
^^ Basically coming from the perspective that, if the L7LB opens the connection to a service VIP, then we must route it to bpf_host to call that program and tail-call here and handle correctly.

If L7LB actually initiates a connection to the backend pod IP directly, then that should go via bpf_host _iff_ `endpoint-routes` mode is enabled. Otherwise it would be routed directly to the destination pod or potentially even off-node. So if this is a possibility, may be worth adding a check in userspace to limit endpointRoutes mode with this feature until we figure out that path & also test it.

joestringer replied:
Aha, second to last commit looks like it addresses this at least for backend IPs :-)

My question remains about whether L7LB typically ends up initiating the questions to VIPs or backends and if VIPs, how to ensure we tail call back to the right place.

jrajahalme replied:
Initially I did all the testing on tunnel mode, so packets trivially ended up in `cilium_host/cilium_net` unless they were for a local backend. That's why the additional tail calls are in a separate commit, plan to squash them before merging...

If the L7LB initiates a connection to a VIP (e.g., service's cluster IP), it will be load balanced to a backend by regular Cilium LB (e.g., in `bpf_sock.c`). We kind of have to do that since we can't do (L3/4) policy enforcement on cluster IPs. So that same call sites that take care of the tail calls for the case where L7LB balances to the backends directly should apply :-)

jrajahalme replied:
Now squashed the nodeport fixup commit together with the main one, hopefully this is all clearer now :-)

joestringer commented on pkg/maps/policymap/callmap.go:
```diff
@@ -72,7 +86,15 @@ func OpenCallMap() (*PolicyPlumbingMap, error) {
 	return fmt.Sprintf("1/%#04x", id)
 }
+
+// EgerssCallString returns the string which indicates the calls map by index in the
```
```suggestion
// EgressCallString returns the string which indicates the calls map by index in the
```

jrajahalme replied:
Fixed, thanks!

joestringer commented on bpf/bpf_sock.c:
```diff
@@ -382,6 +385,35 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
+		 * but we need to redirect for L7 LB also in that case.
+		 */
+		if (is_defined(BPF_HAVE_NETNS_COOKIE) && in_hostns) {
+			/* Use the L7 LB proxy port as a backend. Normally this
+			 * would cause policy enforcement to be done before the
+			 * L7 LB (which should not be done), but in this case
+			 * (node-local nodeport) there is no policy enforcement
+			 * anyway.
+			 */
+			l7backend.address = bpf_htonl(0x7f000001);
+			l7backend.port = (__be16)svc->l7_lb_proxy_port;
```
Is there one of these per nodeport or something? Or, how does the L7LB know which service is being contacted?

jrajahalme replied:
TL;DR: Typically yes.

My initial design was to try always preserve the service IP in the destination address so that the Envoy listener could make a difference between them if multiple `CiliumEnvoyConfig.Spec.Services` are defined. However, in this case of node-port from the host networking namespace there is no bpf program on the TC level where we could redirect the traffic. And in this case the destination IP is the node address anyway, rather than the service IP.

It looks like doing this in most cases would allow us to simplify the TC level handling, basically getting rid of the per-packet lb stuff there. Given this I'd be inclined to document that the destination IP address as seen by the Envoy Listener can be the local host address, node address, as well as the actual service IP.

At least for K8s Ingress implementation this does not matter as there is only one `CEC.Spec.Services` and each Ingress gets their own Envoy listener in a separate listening port.

brb replied:
Where in the agent code do we enforce that there is a single l7 lb proxy port per svc?

(see comment below -- jarno)

jrajahalme replied:
https://github.com/cilium/cilium/pull/18894/files#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672R246-R253

jrajahalme replied:
https://github.com/cilium/cilium/pull/18894/files#diff-2b655ba613ca8386fd9f4c383200480492ee25711e24a575f3fc8b33531b4672R246-R253

joestringer commented on bpf/bpf_sock.c:
```diff
@@ -382,6 +385,35 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
+		/* TC level eBPF datapath does not handle node local traffic,
+		 * but we need to redirect for L7 LB also in that case.
+		 */
+		if (is_defined(BPF_HAVE_NETNS_COOKIE) && in_hostns) {
```
The commit here says that:

> This commit implements L7 LB redirection for nodeport services received by the node from other nodes or from an external load balancer.

But this is the socket hook, and we're checking that the connection is being established from the root netns...so this is only for the *source* node, ie "for nodeport services initiated to remote nodes from this node" right? How could connections from outside the node hit the `connect()` / `sendmsg()` socket hooks here?

jrajahalme replied:
They can't, I'll have to adjust the commit message (done).

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -1729,6 +1739,15 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+			send_trace_notify(ctx, TRACE_TO_PROXY, src_identity, 0,
+					  bpf_ntohs((__u16)svc->l7_lb_proxy_port), 0,
+					  TRACE_REASON_POLICY, monitor);
+			return ctx_redirect_to_proxy_hairpin_ipv4(ctx,
```
Which device do we expect this hairpinning to occur on? Wouldn't it be the native device? In that case, how do we ensure that the traffic going back out the incoming device gets somehow redirected to the proxy rather than just leaving the node?

jrajahalme replied:
Good guestion, and this may require further scrutiny, as I arrived to the use of "hairpin" experimentally; The datapath did not work without the "hairpin" and works with it...

brb replied:
> Wouldn't it be the native device

How is it possible if we redirect in the `ctx_redirect_to_proxy_hairpin()` to `HOST_IFINDEX` which is `cilium_net`?

joestringer reviewed (commented):
I took an initial look, just over the datapath commits. I skipped the big refactor one `4f2a4999e51a datapath: Reduce from LXC complexity` so far, but reviewed all of the rest.

joestringer commented on bpf/lib/dbg.h:
```diff
@@ -119,6 +119,7 @@ enum {
 	DBG_SK_ASSIGN,		/* arg1: result
 				 * arg2: unuseds
 				 */
+	DBG_L7_LB,
```
Please add the args here in comments so readers can interpret the fields and double-check correctness.

jrajahalme replied:
```
-	DBG_L7_LB,
+	DBG_L7_LB,		/* arg1: saddr (last 4 bytes for IPv6)
+				 * arg2: daddr (last 4 bytes for IPv6)
+				 * arg3: proxy port (in host byte order)
+				 */
```
Not sure if using the same enum value for IPv4 and IPv6 works out?

joestringer commented on bpf/lib/common.h:
```diff
@@ -482,6 +484,7 @@ enum metric_dir {
  *    In the IPsec case this becomes the SPI on the wire.
  */
 #define MARK_MAGIC_HOST_MASK		0x0F00
+#define MARK_MAGIC_PROXY_EGRESS_EPID	0x0900 /* mark carries source endpoint ID */
```
Can you also update the text above to highlight that depending on this field, the upper 16 bits hold endpoint ID instead?

jrajahalme replied:
Done.

joestringer commented on install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml:
```diff
@@ -102,6 +102,9 @@ rules:
   - ciliumendpointslices
+  - ciliumenvoyconfigs
+  - ciliumenvoyconfigs/status
+  - ciliumenvoyconfigs/finalizers
```
Do you need finalizers?

joestringer replied:
Ditto for other clusterroles, if you're not using these, please remove them.

jrajahalme replied:
Removed /status and /finalizers as neither is (currenly) used.

joestringer reviewed (commented):
Couple of other notes while glancing through my codeowners.

ldelossa commented on pkg/service/service.go:
```diff
@@ -309,13 +445,66 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
+	name := svc.svcNamespace + "/" + svc.svcName
+	lbEndpoints := []*envoy_config_endpoint.LbEndpoint{}
+	for _, be := range svc.backends {
+		lbEndpoints = append(lbEndpoints, &envoy_config_endpoint.LbEndpoint{
```
This is getting into "curly bracket hell". 

Can you define a variables for these structures ahead of time and then place the vars in this data structure you're defining?

jrajahalme replied:
Here the innermost struct carries the port number that is unique for each instance. If we defined this as a variable we'd need to take a deep copy due to every member in the protobuf being a pointer, and then modify the port value inside. Not sure if that would be better?

ldelossa commented on pkg/service/service.go:
```diff
@@ -309,13 +445,66 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
+	var resources envoy.Resources
+	resources.Endpoints = append(resources.Endpoints, endpoint)
+	if s.monitorNotify != nil {
+		return s.monitorNotify.UpsertEnvoyResources(context.TODO(), resources, nil)
```
Will we ship with context.TODO() ?

jrajahalme replied:
In this case we can, as the context is only used when upserting listener resources, which is not the case here. I'll add a comment to this effect.

ldelossa reviewed (approved):
Took a look at the watcher code and the service implementation. 

Couple comments but nothing to block the PR over.

aditighag commented on pkg/service/service.go:
```diff
@@ -141,6 +144,8 @@ type Service struct {
 	lbmap         LBMap
 	lastUpdatedTs atomic.Value
+
+	l7lbSvcs map[string]uint16 // key: namespace/name, value: local proxy port
```
Use `k8s.ServiceID` as the key.

jrajahalme replied:
The service may not yet exist when this map is populated due to a CEC being applied. As such the serviceID would not be available at the time.

aditighag commented on pkg/service/service.go:
```diff
@@ -162,8 +167,13 @@ func NewService(monitorNotify monitorNotify) *Service {
 	svc.lastUpdatedTs.Store(time.Now())
+
+	// XXX: Hard-coded test data
+	svc.l7lbSvcs["cilium-test/echo-other-node"] = uint16(9090)
```
Leftovers from debugging?

jrajahalme replied:
Yes, removed in later commits, will clean up on the second round of reviews.

aditighag commented on pkg/service/service.go:
```diff
@@ -734,7 +755,7 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 
 	// Update sessionAffinity
-	if option.Config.EnableSessionAffinity {
+	if option.Config.EnableSessionAffinity && svc.l7LBProxyPort == 0 {
```
Need more context here - This code block processes session affinity related code. Why is this `svc.l7LBProxyPort == 0` check added here?

jrajahalme replied:
L7 LB shares the same union in the service map value, but as these features are not needed at the same time, L7 LB takes the precedence over session affinity. proxy port is non-zero when L7 LB is registered for this service.

I added utility function `isL7LBService()` and used it here. Will add a comment to the above effect as well.

aditighag commented on pkg/service/service_test.go:
```diff
@@ -861,3 +861,61 @@ func (m *ManagerTestSuite) TestRestoreServiceWithTerminatingBackends(c *C) {
+// l7 load balancer service should be able to override a ClusterIP service with same
+// frontend, but reverse should produce an error. Also, it should not override
+// any other type besides itself or clusterIP type.
+func (m *ManagerTestSuite) TestL7LoadBalancerServiceOverride(c *C) {
```
The test isn't doing what the text above describes. It's adding services of the same type `SVCTypeClusterIP` back-to-back instead of `SVCTypeL7LoadBalancer`? Anyway I don't see any code that tackles this - 

>l7 load balancer service should be able to override a ClusterIP service with same
// frontend, but reverse should produce an error.

jrajahalme replied:
I had neglected to update the comment as the design evolved. It now reads:
```
// l7 load balancer service should be able to override any service type
// (Cluster IP, NodePort, etc.) with same frontend.
```

aditighag reviewed (changes requested):
First pass over pkg/service* and bpf/bpf_sock*.

Can you add some details to the commit description with the motivation and overview of the changes to help the reviewers and for future reference?

>l7 load balancer service should be able to override a ClusterIP service with same
// frontend, but reverse should produce an error.

Where and how is this enforced? I see only 1 commit titled `lb:`, and I don't see any code that tackles this requirement.

brb commented on bpf/bpf_lxc.c:
```diff
@@ -155,6 +155,11 @@ ipv6_l3_from_lxc(struct __ctx_buff *ctx, struct ipv6_ct_tuple *tuple,
 		svc = lb6_lookup_service(&key, is_defined(ENABLE_NODEPORT));
 		if (svc) {
+			if (lb6_svc_is_l7loadbalancer(svc)) {
+				verdict = svc->l7_lb_proxy_port;
```
The code would be easier to read if you had a dedicated var for the proxy port instead of reusing the verdict var.

jrajahalme replied:
`verdict` normally carries the proxy port from the policy enforcement code (if any), so here we reuse `verdict` the same way it is used elsewhere. Later commits make this a bit clearer, I hope:
```
		/* Restore data from service handling. */
#if defined(ENABLE_L7_LB)
		verdict = ctx_load_meta(ctx, CB_PROXY_MAGIC) >> 16;
		if (verdict != 0) {
			proxy_port = (__u16)verdict;
			/* tuple addresses have been swapped by CT lookup */
			cilium_dbg3(ctx, DBG_L7_LB, tuple.daddr.p4, tuple.saddr.p4,
				    bpf_ntohs(proxy_port));
			ctx_store_meta(ctx, CB_PROXY_MAGIC, 0);
			goto skip_policy_enforcement;
		}
#endif /* ENABLE_L7_LB */
```
Even in this case the code after `skip_policy_enforcement` label will use the `verdict` variable like it needs to if we did not skip policy enforcement. Here passing the proxy port in `verdict` is a method to reuse the redirection logic that is otherwise used for L7 egress policy enforcement.

brb commented on pkg/service/service.go:
```diff
@@ -319,6 +327,43 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
+	var resources envoy.Resources
+	resources.Endpoints = append(resources.Endpoints, endpoint)
+	if s.monitorNotify != nil {
+		return s.monitorNotify.UpsertEnvoyResources(context.TODO(), resources, false)
```
Maybe a bit of code smell. I think the `monitorNotify` was introduced to send updates to the monitoring subsystem (i.e., Hubble). In this case, it's used to update xDS resources which is non-related to the former subsystem. Have you considered calling it instead from a place in which `(s *Service).UpsertService()` is invoked?

jrajahalme replied:
Done by adding a new `envoyCache` interface with only the functions that were actually used. Also rearranged daemon bootstrap a bit so that `d.l7Proxy` is initialized (if enabled) before the `NewService()` call.

brb commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumenvoyconfigs.yaml:
```diff
@@ -0,0 +1,115 @@
+            type: object
+          spec:
+            properties:
+              backendServices:
```
It took me awhile to understand what `backendServices` is and how it differs from `services`. I find the name a bit misleading. Maybe `servicesWhichBypassEnvoyLB` or so could be a better term.

jrajahalme replied:
All services bypass L7 LBs by default. I'm open for better names in the CRD API, but it would need to be more like `servicesWhoseBackendsAreSyncedToEnvoyWithoutFrontendBeingRedirected` I have some follow-on work queued up in this area, so we can come back to naming also after this one merges.

JAORMX replied:
To be honest, I still struggle to understand the name. The first time I saw it I couldn't figure out what this was and the description wasn't very descriptive either.  Perhaps a better explanation could be given in the description? But ideally another name too.

jrajahalme replied:
Updated the description:
```
 	// BackendServices specifies Kubernetes services whose backends
-	// are automatically synced to Envoy usign EDS.
+	// are automatically synced to Envoy using EDS.  Traffic for these
+	// services is not forwarded to an Envoy listener. This allows an
+	// Envoy listener load balance traffic to these backends while
+	// normal Cilium service load balancing takes care of balancing
+	// traffic for these services at the same time.
```

jrajahalme replied:
I did not come up with a better name yet.

JAORMX replied:
thanks fine, the description addition is already very helpful, thank you!

brb commented on pkg/service/service.go:
```diff
@@ -847,7 +899,7 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 	}
 
-	if option.Config.EnableSessionAffinity {
+	if option.Config.EnableSessionAffinity && svc.l7LBProxyPort == 0 {
```
Please add `(s *svc) IsL7Service()` and call it here.

jrajahalme replied:
Done :-)

brb commented on pkg/service/service.go:
```diff
@@ -396,6 +441,13 @@ func (s *Service) UpsertService(params *lb.SVC) (bool, lb.ID, error) {
 		return false, lb.ID(0), err
 	}
 
+	if svc.l7LBProxyPort != 0 {
```
Ditto as bellow.

jrajahalme replied:
Done.

kaworu commented on pkg/hubble/parser/seven/parser_test.go:
```diff
@@ -72,7 +72,7 @@ func TestDecodeL7HTTPRecord(t *testing.T) {
 		DropReason:          nil,
 		HTTP: &accesslog.LogRecordHTTP{
-			Code:     404,
+			Code:     0,
```
nitpicking: Should we leave out as effectively it will be the zero value? I find the explicit Code=0 slightly confusing but I could be missing something.

```suggestion
```

jrajahalme replied:
IMO it is OK to have this here for documentation purposes even when 0 is the default value. The other thing is that this change is pulled here from https://github.com/cilium/cilium/pull/18783, so we'll manage reviews for it there.

Sorry for the confusion :-)

kaworu commented on pkg/hubble/parser/seven/parser_test.go:
```diff
@@ -126,6 +126,105 @@ func TestDecodeL7HTTPRecord(t *testing.T) {
+	assert.Equal(t, pb.Verdict_FORWARDED, f.GetVerdict())
+
+	assert.Equal(t, &pb.HTTP{
+		Code:     0,
```
nitpicking: ditto

```suggestion
```

jrajahalme replied:
Same :-)

kaworu reviewed (approved):
Helm related changes LGTM. I'm missing context about the Hubble changes, but it seems to me that the patch has already been ack'd by @glibsm (see [this PR](https://github.com/cilium/cilium/pull/18783)).

brb commented on pkg/datapath/linux/config/config.go:
```diff
@@ -245,7 +245,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 			cDefinesMap["ENABLE_HOST_SERVICES_UDP"] = "1"
 		}
-		if option.Config.EnableHostServicesTCP && option.Config.EnableHostServicesUDP && !option.Config.BPFSocketLBHostnsOnly {
+		if option.Config.EnableHostServicesTCP && option.Config.EnableHostServicesUDP && !option.Config.BPFSocketLBHostnsOnly && !option.Config.EnableEnvoyConfig {
```
Nit: please add a comment explaining why we do per-packet LB in the case of `EnableEnvoyConfig`.

jrajahalme replied:
Will do, but now I see that there is no reason preventing setting `ENABLE_HOST_SERVICES_FULL` if `enable-envoy-config` is set. We'll define `ENABLE_PER_PACKET_LB` if `ENABLE_L7_LB` is set regardless of the value of `ENABLE_HOST_SERVICES_FULL`:
```
#if !defined(ENABLE_HOST_SERVICES_FULL) || \
    defined(ENABLE_SOCKET_LB_HOST_ONLY) || \
    defined(ENABLE_L7_LB)
# define ENABLE_PER_PACKET_LB 1
#endif
```
I'll add the comment above this code in `bpf_lxc.c`.

jrajahalme replied:
Edit: the change you quoted above was reverted in the next commit, so I'm only adding the comment..

jrajahalme replied:
Added this comment:
```
/* Per-packet LB is needed if all LB cases can not be handled in bpf_sock.
 * Most services with L7 LB flag can not be redirected to their proxy port
 * in bpf_sock, so we must check for those via per packet LB as well.
 */
#if !defined(ENABLE_HOST_SERVICES_FULL) || \
    defined(ENABLE_SOCKET_LB_HOST_ONLY) || \
    defined(ENABLE_L7_LB)
# define ENABLE_PER_PACKET_LB 1
#endif
```

brb commented on bpf/lib/lb.h:
```diff
@@ -1058,10 +1078,10 @@ struct lb4_service *lb4_lookup_service(struct lb4_key *key,
 	if (svc) {
 		if (!scope_switch || !lb4_svc_is_local_scope(svc))
-			return svc->count ? svc : NULL;
+			return (svc->count || lb4_svc_is_l7loadbalancer(svc)) ? svc : NULL;
```
Nit: a small comment why it's fine to return a svc with zero endpoints in the case of L7 LB would be useful for future readers.

jrajahalme replied:
Added this comment:
```
			/* Packets for L7 LB are redirected even when there are no backends. */
```

brb commented on bpf/bpf_host.c:
```diff
@@ -478,7 +478,7 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 		if (ctx_get_xfer(ctx) != XFER_PKT_NO_SVC &&
 		    !bpf_skip_nodeport(ctx)) {
 			ret = nodeport_lb4(ctx, secctx);
-			if (ret < 0)
```
Nit: a comment in what / for what case we might expect `TC_ACT_REDIRECT` as the return.

jrajahalme replied:
Added this:
```
			/* nodeport_lb4() returns with TC_ACT_REDIRECT for
			 * traffic to L7 LB. Policy enforcement needs to take
			 * place after L7 LB has processed the packet, so we
			 * return to stack immediately here with
			 * TC_ACT_REDIRECT.
			 */
```
Added similar comment to the IPv6 case as well.

brb commented on pkg/service/service.go:
```diff
@@ -160,13 +195,122 @@ func NewService(monitorNotify monitorNotify) *Service {
 
+// RegisterL7LBService makes the given service to be locally forwarded to the
+// given proxy port, if non-zero, or for backend sync only, if zero.
+func (s *Service) RegisterL7LBService(name, namespace string, cecName string, proxyPort uint16) error {
```
I had a bit of back and forth with the changes in `pkg/service`. I was wondering whether it would not be possible for the envoy config watcher to call the `UpsertService()` method, and move all envoy related bits from the `pkg/service` to the watcher or so?

My main concern is that we are breaking the separation of concerns wrt the service handling. The intention of the `pkg/service` is just to update the lbmap BPF maps. With this PR we are leaking part of envoy service management logic into the package.

I wonder why e.g. `upsertEnvoyEndpoints()` cannot be invoked from the envoy-related watcher?

jrajahalme replied:
CiliumEnvoyConfig watcher does not know when endpoints are added/removed, but I managed to move all this Envoy specific code to `pkg/envoy` by importing `pkg/service` from there.

brb replied:
I have the following suggestions:
1) Invoke the registration of service endpoints via `pkg/envoy` in https://github.com/cilium/cilium/blob/1381542cf12c1438c2f458cc4db4e83455c30c78/pkg/k8s/watchers/watcher.go#L822 (in this place we know endpoints).
2) Replace `RegisterL7LBService()` with `SetL7LB()`, and call it from `pkg/k8s/watchers/cilium_envoy_config.go` instead.

These would allow to eliminate any Envoy related bits from pkg/service, and thus would address my concern:
> My main concern is that we are breaking the separation of concerns wrt the service handling. The intention of the pkg/service is just to update the lbmap BPF maps. With this PR we are leaking part of envoy service management logic into the package.

jrajahalme replied:
I already addressed this concern as I indicated above:
> CiliumEnvoyConfig watcher does not know when endpoints are added/removed, but I managed to move all this Envoy specific code to pkg/envoy by importing pkg/service from there.

There is no dependency from `pkg/service` to Envoy any more:
```
service % git grep "pkg/envoy"
service % 
```

brb replied:
> There is no dependency from pkg/service to Envoy any more:

I'm talking about `envoyCache` and friends in `pkg/service/service.go`.

jrajahalme replied:
Could you elaborate a bit on your suggestion above. I think `pkg/service` needs to be the source of the "backend changed trigger", are you saying `pkg/service` should call to `pkg/k8s/watchers` whenever any backend of any service is added/updated/removed?

Currently `pkg/service` is not importing anything from `pkg/k8s`, adding that might create an import loop.

jrajahalme replied:
So my logic was to add enough state to `pkg/service` to be able to trigger envoy endpoint updates when needed, using an interface to avoid direct dependency.

Would adding a new file in `pkg/service` with all (or most) of the new state address your concern?

jrajahalme replied:
Or are you saying that `pkg/k8s/watchers` intrinsically knows all service backend adds/updates/deletes?

brb replied:
> Or are you saying that pkg/k8s/watchers intrinsically knows all service backend adds/updates/deletes?

Discussed offline. Just for others - yes, watchers know service backends.

brb reviewed (changes requested):
Did the 2nd round of reviewing. Do we plan to add CI tests for this feature?

aditighag commented on pkg/service/service.go:
```diff
@@ -138,13 +169,17 @@ type Service struct {
 	lastUpdatedTs atomic.Value
+
+	// key: namespace/name
+	l7lbSvcs map[string]*L7LBInfo
```
From the previous [comment](https://github.com/cilium/cilium/pull/18894/#discussion_r814965845) - 

I'm referring to this :) - 
```

type ServiceID struct {
    Name      string `json:"serviceName,omitempty"`
    Namespace string `json:"namespace,omitempty"`
}
```

jrajahalme replied:
This is a `pkg/k8s` concept, and as of now `pkg/service` does not import `pkg/k8s`. `pkg/service` uses `ServiceID` from `pkg/loadbalancer`, which is defined as:
```
loadbalancer/loadbalancer.go-// ServiceID is the service's ID.
loadbalancer/loadbalancer.go:type ServiceID uint16
```

But I see that using a struct of strings would be more efficient than formatting strings for keys, so I'll add a new `ServiceName` type for this in `pkg/service`.

jrajahalme replied:
Done :-)

aditighag reviewed (changes requested):
>l7 load balancer service should be able to override a ClusterIP service with same
// frontend, but reverse should produce an error.

>Where and how is this enforced? I see only 1 commit titled lb:, and I don't see any code that tackles this requirement.

Looks like this is still unresolved?

(see comment below -- jarno)

sayboras commented on daemon/cmd/daemon.go:
```diff
@@ -509,6 +509,18 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+	// daemon's proxy into the k8s watcher and each endpoint.
+	bootstrapStats.proxyStart.Start()
+	// FIXME: Make the port range configurable.
+	if option.Config.EnableL7Proxy {
```
should this condition include new flag as well (e.g. option.EnableEnvoyConfig) ?

jrajahalme replied:
Looked at this, I think we should not start the proxy if `EnableL7Proxy` is `false`, even if `EnableEnvoyConfig` is true. I added a warning if L7 proxies are disabled and Envoy Configs are enabled.

JAORMX commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumenvoyconfigs.yaml:
```diff
@@ -0,0 +1,118 @@
+                  - namespace
+                  type: object
+                type: array
+              ingress:
```
what about exchanging this boolean for `flowtype` with the options being `ingress` and `egress`? perhaps it would be a little more descriptive.

jrajahalme replied:
I'll remove `ingress` for now, as it is not supported yet at all. We'll add it back in some form if there is a use case for it and we have a datapath implementation for it.

JAORMX reviewed (changes requested):
I left a comment on the `backendServices` property of the `CiliumEnvoyConfig` CRD. I couldn't unresolve the comment, so I'll just post the same here:

> To be honest, I still struggle to understand the name. The first time I saw it I couldn't figure out what this was and the description wasn't very descriptive either. Perhaps a better explanation could be given in the description? But ideally another name too.

JAORMX commented on pkg/service/service.go:
```diff
@@ -160,13 +196,122 @@ func NewService(monitorNotify monitorNotify) *Service {
 }
 
+// RegisterL7LBService makes the given service to be locally forwarded to the
+// given proxy port, if non-zero, or for backend sync only, if zero.
```
I think it would be more intuitive to make different calls that represent what's actually going on, instead of having the port being zero or not present different functionality (even if it's quite similar). As a reader, it would make it easier to read and understand if there was a function for registering local forwarding and another one for backend sync only.

jrajahalme replied:
Added a new `RegisterL7LBServiceBackendSync()` that does not take the zero port as an argument.

sayboras commented on pkg/k8s/apis/cilium.io/v2alpha1/cec_types.go:
```diff
@@ -0,0 +1,167 @@
+// UnmarshalJSON ensures that the unstructured object properly decodes
+// JSON when passed to Go's standard JSON library.
+func (u *XDSResource) UnmarshalJSON(b []byte) (err error) {
+	if option.Config.Debug {
```
I think this is useful as part of development and debugging, however do you still need this going forward ?

jrajahalme replied:
I moved this to the end to log the given JSON if unmarshaling fails and debug is enabled. Example output below:
```
level=debug msg="CEC UnmarshalJSON failed for: {\n\t\"@type\": \"type.googleapis.com/envoy.config.cluster.v3.Cluster\",\n\t\"connect_timeout\": \"3s\",\n\t\"lb_policy\": \"ROUND_ROBIN\",\n\t\"name\": \"cilium-test/echo-other-node\",\n\t\"outlier_detection\": {\n\t\t\"consecutive_local_origin_failure\": 2,\n\t\t\"split_external_local_origin_errors\": true\n\t},\n\t\"type\": \"ERDS\"\n}" subsys=k8s-apis-cilium.io-v2alpha1
level=warning msg="github.com/cilium/cilium/pkg/k8s/watchers/cilium_envoy_config.go:81: failed to list *v2alpha1.CiliumEnvoyConfig: proto:\u00a0(line 1:258): invalid value for enum type: \"ERDS\"" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/cilium_envoy_config.go:81: Failed to watch *v2alpha1.CiliumEnvoyConfig: failed to list *v2alpha1.CiliumEnvoyConfig: proto:\u00a0(line 1:258): invalid value for enum type: \"ERDS\"" subsys=k8s
```

sayboras commented on pkg/k8s/watchers/cilium_envoy_config.go:
```diff
@@ -0,0 +1,149 @@
+
+func (k *K8sWatcher) addCiliumEnvoyConfig(cec *cilium_v2alpha1.CiliumEnvoyConfig) error {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.CiliumEnvoyConfigName: cec.ObjectMeta.Name,
```
understand that CEC is logically clusterwide resource, however, it's still having namespace, so we can add namespace here in log.

jrajahalme replied:
CEC it not namespaced, so we can't add namespace here. Referred services, if any, are namespaced, but can not be added here, as they might not exist. Services can be from different namespaces, so not sure what to log.

But we should discuss, whether CEC should be namespaced to begin with. Listener ports are created in the host networking namespace, so not sure what namespacing CECs would mean. Maybe we would then restrict all referred services to the namespace of the CEC itself?

sayboras commented on pkg/envoy/ciliumenvoyconfig.go:
```diff
@@ -0,0 +1,480 @@
+					Address: &envoy_config_core.Address{
+						Address: &envoy_config_core.Address_SocketAddress{
+							SocketAddress: &envoy_config_core.SocketAddress{
+								Address: be.L3n4Addr.IP.String(),
```
do we need to add Protocol from be.L3n4Addr.L4Addr.Protocol ?

jrajahalme replied:
CECs only support TCP (at least for now, which is the default for Envoy `SocketAddress`. I need to filter non-TCP backends away in here... Done.

sayboras commented on pkg/k8s/watchers/cilium_envoy_config.go:
```diff
@@ -0,0 +1,149 @@
+		scopedLog.WithError(err).Warn("Failed to add CiliumEnvoyConfig: malformed Envoy config.")
+		return err
+	}
+	if err := k.envoyConfigManager.UpsertEnvoyResources(context.TODO(), resources, k.envoyConfigManager); err != nil {
```
nit: the caller is k.envoyConfigManager, and one of param is also k.envoyConfigManager, which is a little bit confusing to me at first.

jrajahalme replied:
The compiler will implicitly cast `k.envoyConfigManager` to `envoy.PortAllocator` on this call. `UpsertEnvoyResources()` needs an `envoy.PortAllocator` as an argument, and it was simplest to merge its methods to `envoyConfigMananger`, rather than passing two different interfaces to the k8s watcher.

sayboras commented on pkg/k8s/watchers/cilium_envoy_config.go:
```diff
@@ -0,0 +1,281 @@
+	}
+	for _, oldSvc := range removedServices {
+		// Tell service manager to remove old service registration
+		err := k.svcManager.RemoveL7LBService(oldSvc.Name, oldSvc.Namespace, oldCEC.ObjectMeta.Name)
```
do you know if this is grateful removal (i.e. inflight requests should be allowed till completion) ?

Also, should we update and then delete ? just to avoid any disruption?

jrajahalme replied:
Good question. This is executed after a CEC is deleted, or updated to not contain a specific service, either in `services` or `backendServices`. After this new connections will no longer be redirected to the Envoy listener, but same will likely happen to the remaining packets of the old connections, as we do not store the proxy port to the conntrack entry.

Similar disruption happens if a non-L7 policy is changed to an L7-policy or the other way around.

We probably should store the proxy port in to the conntrack entry just like the `backend_id` is for the normal Cilium LB.

sayboras reviewed (commented):
It will take some more time for me to understand two datapath commits, so I just review other commits first.

jrajahalme committed:
envoy: Ignore invalid JSON in CiliumEnvoyConfig resources

Ignore invalid JSON in CiliumEnvoyConfig resources in order to not break
k8s resource listers. Ignored resources are logged on warning level. The
rest of the CEC is applied and may or may not function depending on the
role of the ignored resource.

Fixes: https://github.com/cilium/cilium-service-mesh-beta/issues/27
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
k8s: Add CiliumClusterwideEnvoyConfig

Change CiliumEnvoyConfig to be namespace scoped, and add a cluster scoped
CiliumClusterwideEnvoyConfig (CCEC).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras reviewed (approved):
LGTM :100:

aanm commented on install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml:
```diff
@@ -98,6 +98,8 @@ rules:
   - ciliumlocalredirectpolicies/finalizers
   - ciliumegressnatpolicies
   - ciliumendpointslices
+  - ciliumclusterwideenvoyconfigs
+  - ciliumenvoyconfigs
```
I know this is current with verbs '*' but is currently being addressed in https://github.com/cilium/cilium/pull/19074

Please create add the specific verbs we will be using in Cilium's ClusterRole instead of `'*'`. Either make the change here or leave a comment in #19074 which verbs will be used for the ciliumenvoyconfigs.

jrajahalme replied:
Changed to only include verbs `list` and `watch`.

aanm commented on pkg/k8s/factory_functions.go:
```diff
@@ -957,3 +1003,47 @@ func ConvertCoreCiliumEndpointToTypesCiliumEndpoint(ccep *cilium_v2alpha1.CoreCi
+		}
+	}
+	log.WithField(logfields.Object, logfields.Repr(obj)).
+		Warn("Ignoring invalid v2 Cilium Envoy Config")
```
`Ignoring invalid v2alpha1 Cilium Clusterwide Envoy Config`

jrajahalme replied:
done.

aanm commented on pkg/k8s/factory_functions.go:
```diff
@@ -957,3 +1003,47 @@ func ConvertCoreCiliumEndpointToTypesCiliumEndpoint(ccep *cilium_v2alpha1.CoreCi
+		}
+	}
+	log.WithField(logfields.Object, logfields.Repr(obj)).
+		Warn("Ignoring invalid v2 Cilium Envoy Config")
```
`Ignoring invalid v2alpha1 Cilium Envoy Config`

jrajahalme replied:
done.

aanm commented on pkg/k8s/watchers/cilium_envoy_config.go:
```diff
@@ -0,0 +1,316 @@
+		}
+	}
+	return service.Name{Name: name, Namespace: namespace}
+}
```
Is this the behavior that we want? My understanding is that users can specify any namespace in the `svc.Namespace` which can be a different namespace than `resourceName.Namespace`? Shouldn't they be the same? For example, in CNP we forcefully set the namespace in `fromEndpoints` to be the same namespace where the CNP is deployed even if the user manually sets a namespace different than the one where CNP is deployed.

jrajahalme replied:
@aanm You tell me, I can tie this down like you suggested, clusterwide variant can then be used for mixed namespace cases if so desired.

aanm replied:
I think it would be better to have the same behavior as we have in CNP. Sorry I just realized I made a mistake. The `endpointSelector` of a CNP is the one that is forcefully set to the same namespace where the CNP is set but the `fromEndpoints` can allow from any namespace. It's defined [here](https://github.com/cilium/cilium/blob/314ca7baeff4f568ffc0bad95124e40665b1f88c/pkg/k8s/apis/cilium.io/utils/utils.go#L298-L316)

jrajahalme replied:
So we keep as then?

aanm replied:
We should have the same behavior as the in the CNPs i.e., we can select from any namespace but we can only _enforce_ ingress on the namespace where cilium envoy config is.

jrajahalme replied:
For Cilium Envoy Configs all Envoy listeners are on egress. I take the above to read that the redirected services MUST be in the same namespace as the CRD itself, so that service access in other namespaces is not affected. Services actually accessed by the L7 LB can be in any namespaces (unless blocked by a (Cilium) Network Policy. Sound about right?

aanm commented on pkg/maps/policymap/callmap.go:
```diff
@@ -44,23 +45,36 @@ func (v *PlumbingValue) String() string {
+		if err == nil {
+			return err2
+		}
+	}
```
Sorry, can we move this out into the caller instead? It looks ugly having to import `github.com/cilium/cilium/pkg/option` just for this use case.

jrajahalme replied:
This done as well, forgot to mention, sorry.

aanm commented on pkg/maps/policymap/policymap.go:
```diff
@@ -441,5 +446,21 @@ func InitCallMap() error {
+
+		_, err = policyEgressCallMap.Create()
+	}
 	return err
```
Same here, let's try and move this to the caller instead.

jrajahalme replied:
You mean passing in a bool for `option.Config.EnableEnvoyConfig`??

aanm replied:
Either that or pass an interface something like it's being done [here](https://github.com/cilium/cilium/blob/f0465215aec02e7c21887d672840f6ff03780374/pkg/service/store/store.go#L162-L188)

jrajahalme replied:
Done.

aanm commented on pkg/service/service.go:
```diff
@@ -60,6 +60,22 @@ type monitorNotify interface {
 
+// Name represents the fully-qualified reference to the service by name, including both the
+// namespace and name of the service.
+type Name struct {
```
`type Name struct` looks too broad. Maybe `type ServiceName struct` instead?

jrajahalme replied:
I had it like that, but then when using from other packages `service.Name` felt much better than `service.ServiceName`. What do you think?

aanm replied:
Sounds good

jrajahalme replied:
So kept as-is.

aanm reviewed (changes requested):
Only a couple of small nits and observations. I don't think it requires a second review from my side.

sayboras commented on pkg/k8s/apis/cilium.io/v2alpha1/cec_types.go:
```diff
@@ -20,7 +20,7 @@ import (
 // +genclient
 // +genclient:nonNamespaced
```
should we change this to namespaced by default as well ?

jrajahalme replied:
Fixed, thanks!

brb reviewed (changes requested):
Comment re pkg/service - https://github.com/cilium/cilium/pull/18894#discussion_r840633249

brb reviewed (approved):
> Comment re pkg/service - https://github.com/cilium/cilium/pull/18894#discussion_r840633249

Discussed offline. This will be addressed in a follow-up after we have figured out a correct path.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -212,7 +212,13 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host)
 		if (ctx_get_xfer(ctx) != XFER_PKT_NO_SVC &&
 		    !bpf_skip_nodeport(ctx)) {
 			ret = nodeport_lb6(ctx, secctx);
-			if (ret < 0)
+			/* nodeport_lb6() returns with TC_ACT_REDIRECT for
+			 * traffic to L7 LB. Policy enforcement needs to take
+			 * place after L7 LB has processed the packet, so we
+			 * return to stack immediately here with
+			 * TC_ACT_REDIRECT.
+			 */
+			if (ret < 0 || ret == TC_ACT_REDIRECT)
 				return ret;
 		}
```
I believe we need the same check in `bpf_overlay.c`'s nodeport code:
https://github.com/cilium/cilium/blob/34d128aa85c61a41175c92b64afacc05bcd615a5/bpf/bpf_overlay.c#L57-L65

joestringer replied:
Was this addressed?
